File: 02-UML-Class-Diagrams/src/java/chapter02/DigitalBook.java
package chapter02;

/**
 * UML: Inheritance.
 * DigitalBook --|> Book  (DigitalBook IS-A Book)
 */
public class DigitalBook extends Book {
    private String format;     // PDF, EPUB, etc.
    private double fileSizeMb;

    public DigitalBook(String title, String isbn, Author author, String format, double fileSizeMb) {
        super(title, isbn, author);
        this.format = format;
        this.fileSizeMb = fileSizeMb;
    }

    public String getFormat() { return format; }
    public double getFileSizeMb() { return fileSizeMb; }

    @Override
    public void displayInfo() {
        System.out.println("[DigitalBook] " + getTitle() + " by " + getAuthor().getName()
                + " | Format: " + format + " | Size: " + fileSizeMb + "MB");
    }
}

File: 02-UML-Class-Diagrams/src/go/digital_book.go
package main

import "fmt"

// Inheritance: Go doesn't have inheritance — use embedding (composition).
// DigitalBook embeds Book and adds extra fields.

type DigitalBook struct {
	*Book              // Embedding = promoted fields/methods (like inheritance)
	Format     string
	FileSizeMb float64
}

func NewDigitalBook(title, isbn string, author *Author, format string, fileSizeMb float64) *DigitalBook {
	return &DigitalBook{
		Book:       NewBook(title, isbn, author),
		Format:     format,
		FileSizeMb: fileSizeMb,
	}
}

// Override DisplayInfo — Go dispatches to this if called on *DigitalBook
func (d *DigitalBook) DisplayInfo() {
	fmt.Printf("[DigitalBook] %s by %s | Format: %s | Size: %.1fMB\n",
		d.Title, d.Author.Name, d.Format, d.FileSizeMb)
}
